### PR TITLE
MongoDB based EnvironmentRepository implementation.

### PR DESCRIPTION
**Pull request by venilnoronha**
Hi,

I've implemented a basic MongoDB based EnvironmentRepository (Issue #203). I've also added a simple test. 

### Configuration Properties
spring.cloud.config.server.mongodb.host - _The mongodb host (optional) (default 127.0.0.1)_
spring.cloud.config.server.mongodb.port - _The mongodb port (optional) (default 27017)_
spring.cloud.config.server.mongodb.database - _The mongodb database_
spring.cloud.config.server.mongodb.username - _The mongodb username (optional)_
spring.cloud.config.server.mongodb.password - _The mongodb password (optional)_

### Enabling MongoEnvironmentRepository
The _spring.profiles.active_ must be set to _'mongodb'_ and at least the _spring.cloud.config.server.mongodb.database_ property must be configured to enable this repository.

Due to some issues I had to close my previous pull requests. Apologies.

Please review and pull.

Thanks,
Venil

### CODE REVIEW

**Comment by venilnoronha** [comment]
Travis is ignoring my MongoEnvironmentRepositoryIntegrationTests class. Its also showing the following output.

> Warning: you are leaving 2 commits behind, not connected to any of your branches:
> 
>   479c869 Merge 811a58addc1a5adf369c7fc621887742da42f0ad into 1c69c59705e4e2e7bf22d511e0b777b75f094c42
> 
>   811a58a MongoDB based EnvironmentRepository implementation. #203.
> 
> If you want to keep them by creating a new branch, this may be a good time to do so with:
> 
>  git branch new_branch_name 479c869

Any idea why?

**Comment by dsyer** [comment]
Why would we need `spring.cloud.*` configuration for Mongo (it's already provided by Spring Boot)?

**Comment by venilnoronha** [comment]
If at all one needs to use a different mongo instance for just storing configs.

**Comment by dsyer** [comment]
Sorry, I didn't catch that. Why would you need a cloud-specific config?

**Comment by venilnoronha** [comment]
I mean, if the config server is to be clubbed with a component already working on a mongo instance, then the spring.cloud.config.server.mongodb.* properties would be useful to configure a separate mongo instance for only serving configuration. (This mongo instance may as well be deployed on another server with a different IP.)

If this usecase doesn't seem to be valid then I'll remove those properties and use the spring-boot mongo instance instead.

**Comment by dsyer** [comment]
I don't think it's a common enough use case that we'd want to support it out of the box with configuration properties. If you can just use the default autowired Mongo client that would help I think. Also, did you fill in the contributor's agreement (link in README)?

**Comment by venilnoronha** [comment]
OK. I'll remove those properties. And yes, I've signed the ICLA. Thanks.

**Comment by venilnoronha** [comment]
I've made the necessary changes to use the spring-boot provided MongoTemplate instance in MongoEnvironmentRepository. Could you please verify and pull?

File: spring-cloud-config-server/pom.xml [hunk 1]
```diff
@@ -56,6 +56,10 @@
 			<optional>true</optional>
 		</dependency>
 		<dependency>
+			<groupId>org.springframework.data</groupId>
+			<artifactId>spring-data-mongodb</artifactId>
+		</dependency>
```
Comment (spencergibb) [inline spring-cloud-config-server/pom.xml]: Should this be optional?

File: spring-cloud-config-server/src/main/java/org/springframework/cloud/config/server/config/EnvironmentRepositoryConfiguration.java [hunk 2]
```diff
@@ -95,5 +99,30 @@ public EnvironmentRepository environmentRepository() {
 			return repository;
 		}
 	}
+	
+	@Configuration
+	@Profile("mongodb")
+	protected static class MongoRepositoryConfiguration {
```
Comment (spencergibb) [inline spring-cloud-config-server/src/main/java/org/springframework/cloud/config/server/config/EnvironmentRepositoryConfiguration.java]: ConditionalOnClass(MongoTemplate.class) to go with optional dependency above.

**Comment by venilnoronha** [comment]
@spencergibb done.

**Comment by DCalata** [comment]
Hey Venil,

We make us of the MongoDBEnvironmentRepository implementation and unfortunately we've found out an issue when it retrieves a 'complex' copnfiguration  (see below the example) which is stored in MongoDB. This problem occurs when we want to obtain the configuration as a property invoking the REST API.

Below an example of this scenario:
 
1)  Mongo and spring cloud configuration running locally on the default ports.

	Configuration stored in mongodb.
	'''   
	   ....
	   properties : {
	        "foo" : {
	            "bar": "fooBar"
	        }
	    }
	    ...
	'''

2)  Then when invoking the following url: http://localhost:8888/master/sample-default.properties

3)   And then the result is the below

		properties: {foo={bar=fooBar}}

	 but we expected a response like the below:

		proterties : foo.bar = fooBar

The reason We expected the aboveis  because when we invoke the yaml endpoint the response looks like this:

 		foo:
   			bar: fooBar

We have been working on fixing this issue and it would be greatful If we could contribute to this project with the fix, would you mind If we do a pull request to your repository?  If so, we would like also to make some improvements in the tests implementation by using Fongo  instead of assuming you have a mongo instance running on localhost

foo.bar: fooBar

Thanks in advance,
Cheers,

**Comment by venilnoronha** [comment]
Sure @DCalata, I'll have a look at it. You can send a PR to my repository as well.

**Comment by venilnoronha** [comment]
@DCalata thanks for highlighting the issue. I've fixed it in 71aee7a.

**Comment by javiboo** [comment]
Hi @venilnoronha,

I'm a partner of @DCalata. I've been testing last changes introduced following their comments. 

I found an issue due `cleanSource` method introduced in b7aa7cb. I make a [test](https://github.com/SerenityDevstack/spring-cloud-config/commit/6efc393439ddfd786be0a85255c254e589323260) in a forked repository that reproduces the issue.

In order to follow the behaviour of the other profiles (native, git & subversion), I introduce this [changes](https://github.com/SerenityDevstack/spring-cloud-config/commit/2aa59ed62f38399064ad063ef27339c5ad47aa62) in the MongoEnvironmentRepository to allow the definition of any key in the name of the properties. Can you check the changes, please? What do you think about it? 

If you consider appropriate, I can make a PR with these changes to your repository so you can amend it to your PR. Thanks.

**Comment by venilnoronha** [comment]
Hi @javiboo,

While I see that the label and profile values are being lost, I'm still wondering what would the use case be for needing those values. And if someone needs them, would there be 2 different profile/labels values per configuration? If not, then we can update `cleanSource` to retain those properties, which would then be returned in the API response.

My concern is that, having 2 label/profile values might cause ambiguity to the user. Also, would it not be nice to avoid nested properties at the first level itself?

Thoughts?

**Comment by javiboo** [comment]
Hi @venilnoronha,

I agree with you and I don't think it's a good idea to have 2 different profile/labels values per configuration.

By defining a property named label, I'm trying to demonstrate the differences between mongo-db profile and the other profiles.

Suppose we have a configuration file `foo.yml` stored in our fs (with _native_ profile active) for an application with name `foo` with the content:
```
property1: value1
label: value2
```
so if you request your configuration through the endpoint `/{name}/{profiles}` (with values _foo_ and _default_, respectively) you obtain the response:
```
{
  "name": "foo",
  "profiles": [
    "default"
  ],
  "label": null,
  "version": null,
  "propertySources": [
    {
      "name": "classpath:/foo.yml",
      "source": {
        "property1": "value1",
        "label": "value2"
      }
    }
  ]
}
```
With mongo-db profile, how can you address the same content without nesting properties under a field and also without restricting the use of certain properties in the configuration file (considering that here the property _label_ is also being used as a configuration property, a part of the meaning of this field in Spring Cloud)?

What do you think about it?

Thanks!

**Comment by venilnoronha** [comment]
Sure @javiboo, I get your point. We can add a `source` field of type `LinkedHashMap<String, Object>` to MongoPropertySource instead of `properties` (just to be consistent with the API response, see below).

```
"propertySources": [
  {
    "name": "classpath:/foo.yml",
    "source": {
      "property1": "value1",
      "label": "value2"
    }
  }
]
```

Regarding the PR, if you'd like to contribute to my repo I'm open for it. However, I'm not quite sure if you might get the attribution for it (as all of the commits of this PR would be squashed). Again, if you'd want to be sure about the attribution, it would be better if you could send in the PR to [spring-cloud/spring-cloud-config](https://github.com/spring-cloud/spring-cloud-config) directly once this PR is merged. I'm also fine with making the changes you've suggested to my repo directly, based on your fork of my repo (if you don't mind).

Thanks,
Venil

**Comment by javiboo** [comment]
@venilnoronha I don't mind, you can make the changes suggested directly to your repo. Thanks!

**Comment by venilnoronha** [comment]
@javiboo I've updated the implementation. :)

**Comment by javiboo** [comment]
@venilnoronha thanks for the quick response. We are very interested that this PR joins the spring-cloud project, we can do something to expedite this issue?

**Comment by venilnoronha** [comment]
@javiboo I'm not too sure about that. Maybe @spencergibb could tell if there are any immediate plans to merge this PR.

**Comment by spencergibb** [comment]
There are not immediate plans. We'll have a discussion about it as a team.

**Comment by venilnoronha** [comment]
@spencergibb sure, thanks!

**Comment by spencergibb** [comment]
@venilnoronha what do you think about hosting this in your user org, or maybe in https://github.com/spring-cloud-incubator. That way we can see how users respond and gauge interest before we decide to move this into core.

**Comment by venilnoronha** [comment]
How do we go about hosting this under spring-cloud-incubator?

**Comment by spencergibb** [comment]
@venilnoronha https://github.com/spring-cloud-incubator/spring-cloud-config-server-mongodb, invite sent.

**Comment by venilnoronha** [comment]
Thanks @spencergibb. I'll commit the code there soon.

**Comment by venilnoronha** [comment]
Snapshot version of spring-cloud-config-server-mongodb is now available on jfrog oss. Check out the github repository [here](https://github.com/spring-cloud-incubator/spring-cloud-config-server-mongodb) for information on maven configuration and usage.

**Comment by spencergibb** [comment]
Closing in favor of https://github.com/spring-cloud-incubator/spring-cloud-config-server-mongodb/commit/ecedae00e22978155606cf17ca1f2c3279769aff

**Pull request closed without merging**